cni-plugin/pkg/install/install.go

```diff
@@ -496,3 +513,84 @@ func setSuidBit(file string) error {
 
 	return nil
 }
+
+// destinationUptoDate compares the given files and returns
+// whether or not the destination file needs to be updated with the
+// contents of the source file.
+func destinationUptoDate(src, dst string) (bool, error) {
+	// Stat the src file.
+	f1info, err := os.Stat(src)
+	if os.IsNotExist(err) {
+		// If the source file doesn't exist, that's an unrecoverable problem.
+		return false, err
+	} else if err != nil {
+		return false, err
+	}
+
+	// Stat the dst file.
+	f2info, err := os.Stat(src)
+	if os.IsNotExist(err) {
+		// If the destination file doesn't exist, it means the
+		// two files are not equal.
+		return false, nil
+	} else if err != nil {
+		return false, err
+	}
+
+	// First, compare the files sizes and modes. No point in comparing
+	// file contents if they differ in size or file mode.
+	if f1info.Size() != f2info.Size() {
+		return false, nil
+	}
+	if f1info.Mode() != f2info.Mode() {
+		return false, nil
+	}
+
+	// Files have the same exact size and mode, check the actual contents.
+	f1, err := os.Open(src)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f1.Close()
+
+	f2, err := os.Open(dst)
+	if err != nil {
+		log.Fatal(err)
+	}
+	defer f2.Close()
+
+	// Iterate the files until we reach the end. If we spot a difference,
+	// we know that the files are not the same. Otherwise, if we reach the
+	// end of the file before seeing a difference, the files are identical.
+	for {
+		b1 := make([]byte, 64000)
+		_, err1 := f1.Read(b1)
```

Comment:
Will `File.Read()` ever return less than a full buffer?  If so, the two reads could get out of sync.

Making a new buffer for each read can be wasteful, usually better to reuse a buffer.

Reply:
I believe it will either fill the buffer or return an error, or both, but shouldn't return less than a full buffer if there is more to read. I'll test this out though.

It does appear that if the EOF is returned, it also fills the read bytes into the buffer so I should compare those in the EOF branch as well.

Reply:
I added some unit tests to make sure this works the way I expect.